File: C++/Greedy/24_k_centers_problem.cpp
/*
Idea:
Recursively exploring all combination of ATMs and combination with least max distance from an ATM.
*/

//{ Driver Code Starts
//Initial Template for C++

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//Back-end complete function Template for C++

class Solution {
public:
    void kCenterHelper(int k, int n, int& result, vector<int>& selected, vector<vector<int>>& mat)
    {
        if (k == 0)
        {
            // Find maximum distance of city from server.
            int maxDist = 0;
            for (int i = 0; i < n; i++)
            {
                int minDistServer = INT_MAX;
                for (int j = 0; j < selected.size(); j++)
                {
                    minDistServer = min(minDistServer, mat[i][selected[j]]);
                }
                maxDist = max(maxDist, minDistServer);
            }
            result = min(result, maxDist);
        }
        else
        {
            int prev = (selected.size() > 0) ? selected.back() : -1;
            for (int i = prev + 1; i < n; i++)
            {
                selected.push_back(i);

                // Recursively find ways of selecting remaining cities.
                kCenterHelper(k - 1, n, result, selected, mat);

                // Backtrack to explore possible way of selection without city 'i'.
                selected.pop_back();
            }
        }
    }

    int selectKcities(int n, int k, vector<vector<int>>& mat)
    {
        vector<int> selected;
        int result = INT_MAX;

        // Recursively try all possible way of selecting 'K' cities.
        kCenterHelper(k, n, result, selected, mat);

        return result;
    }
};


//{ Driver Code Starts.


int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, k;
        cin >> n >> k;
        vector<vector<int>> mat(n, vector<int>(n));
        for (int i = 0;i < n;i++) {
            for (int j = 0;j < n;j++) {
                cin >> mat[i][j];
            }
        }

        Solution obj;
        int ans = obj.selectKcities(n, k, mat);

        cout << ans << endl;
    }
}
// } Driver Code Ends

File: Python/Greedy/24_k_centers_problem.py
"""
Idea:
Recursively exploring all combination of ATMs and combination with least max distance from an ATM.
"""


class Solution:
    def kCenterHelper(self, k, n, result, selected, mat):
        if k == 0:
            # Find the maximum distance of a city from the server.
            maxDist = 0
            for i in range(n):
                minDistServer = float('inf')
                for j in range(len(selected)):
                    minDistServer = min(minDistServer, mat[i][selected[j]])
                maxDist = max(maxDist, minDistServer)
            result[0] = min(result[0], maxDist)
        else:
            prev = selected[-1] if selected else -1
            for i in range(prev + 1, n):
                selected.append(i)

                # Recursively find ways of selecting the remaining cities.
                self.kCenterHelper(k - 1, n, result, selected, mat)

                # Backtrack to explore possible ways of selection without city 'i'.
                selected.pop()

    def selectKcities(self, n, k, mat):
        selected = []
        result = [float('inf')]

        # Recursively try all possible ways of selecting 'K' cities.
        self.kCenterHelper(k, n, result, selected, mat)

        return result[0]


# {
# Driver Code Starts
class IntArray:
    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


class IntMatrix:
    def __init__(self) -> None:
        pass

    def Input(self, n, m):
        matrix = []
        # matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix

    def Print(self, arr):
        for i in arr:
            for j in i:
                print(j, end=" ")
            print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        a = IntArray().Input(2)
        n, k = a[0], a[1]

        mat = IntMatrix().Input(a[0], a[0])

        obj = Solution()
        res = obj.selectKcities(n, k, mat)

        print(res)

# } Driver Code Ends
